### PR TITLE
feat(lint): `ignoreRestSiblings` option to `noUnusedVariables`

### PR DESCRIPTION
**Pull request by iamakulov**
## Summary

This pull request adds an `ignoreRestSiblings` option for `lint/correctness/noUnusedVariables`. The option is similar to its ESLint alternative: https://eslint.org/docs/latest/rules/no-unused-vars#ignorerestsiblings

The option defaults to `true` (**unlike** its ESLint alternative) to avoid introducing a breaking change. (I’d love it if we could flip the default to `false` in the next major – as that would allow stricter linting by default – but this is not my call to make.)

Closes: https://github.com/biomejs/biome/discussions/5152

Note: I have zero experience with Rust, so I had to Cursor my way through this PR. The (new) tests do pass, but sorry if there are any glaring omissions!

## Test Plan

Tests should pass, I guess?

### CODE REVIEW

**Comment by iamakulov** [comment]
Thanks for approving the CI run! I fixed linting/formatting checks. Two jobs (codegen and tests) are still falling. 

#### Codegen
Codegen is obvious – but I can’t run it locally due to a silly error (which I can’t figure out how to fix), so I’d appreciate if you can do it:

```sh
$ cargo codegen-configuration
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.52s
     Running `target/debug/xtask_codegen configuration`
> rustfmt --version
rustfmt 1.8.0
Error: Failed to run rustfmt from toolchain 'stable'. Please run `rustup component add rustfmt --toolchain stable` to install it.

$ rustup component add rustfmt --toolchain stable
info: component 'rustfmt' for target 'aarch64-apple-darwin' is up to date

$ which rustfmt
/opt/homebrew/bin/rustfmt
```

#### Tests
I’m not sure why there’s such a diff between local and CI test runs, but I’m happy to hand-fix the snapshots if you say this is expected:

Screenshot 2025-02-19 at 14 07 23

**Review by ematipico — APPROVED** [review]

File: crates/biome_js_analyze/src/lint/correctness/no_unused_variables.rs [hunk 1]
```diff
@@ -34,6 +59,18 @@ declare_lint_rule! {
     /// enable [noUnusedFunctionParameters](https://biomejs.dev/linter/rules/no-unused-function-parameters/).
     /// :::
     ///
+    /// ## Options
+    ///
```
Comment (ematipico) [inline crates/biome_js_analyze/src/lint/correctness/no_unused_variables.rs:63]:
```suggestion
    /// ## Options
    ///
    /// **Since v2.0.0**
    ///
```

**Review by ematipico — CHANGES_REQUESTED** [review]
Can you please add a changeset?

File: crates/biome_js_analyze/src/lint/correctness/no_unused_variables.rs [hunk 2]
```diff
@@ -14,6 +14,29 @@ use biome_js_syntax::{
...
+pub struct NoUnusedVariablesOptions {
+    /// Whether to ignore unused variables from an object desctructuring with a spread
+    /// (i.e.: whether `a` and `b` in `const { a, b, ...rest } = obj` should be ignored by this rule).
+    #[serde(default = "enabled")]
+    ignore_rest_siblings: bool,
+}
+
+const fn enabled() -> bool {
+    true
+}
```
[the remark this answers is not in the document]
Comment (iamakulov) [reply crates/biome_js_analyze/src/lint/correctness/no_unused_variables.rs:33]: The `enabled` function is also used here:

https://github.com/biomejs/biome/blob/84814b856f4001c8fe196c3dfa666329bc7e0409/crates/biome_js_analyze/src/lint/correctness/no_unused_variables.rs#L27

(I copied this from here:

https://github.com/biomejs/biome/blob/84814b856f4001c8fe196c3dfa666329bc7e0409/crates/biome_js_analyze/src/lint/style/use_filenaming_convention.rs#L403

)

I’m guessing this means it still has to stay, right? (Is there a way to set the default serialization value without defining a function?)

[the remark this answers is not in the document]
Comment (ematipico) [reply crates/biome_js_analyze/src/lint/correctness/no_unused_variables.rs:33]: Remove the `default = "enabled"` for `default`. Since you already applied the trait `Default` to the type you created, serde will be able to pick it up automatically

**Comment by iamakulov** [comment]
> I don't usually use `brew` to handle rust on my machine. Rust has a tool called `rustup` that does everything without problems. https://www.rust-lang.org/tools/install

Thanks! Replacing the brew-installed `rust` and `rustup` with the full version helped. I updated the codegen + addressed the review feedback.

**Comment by Conaclos** [comment]
> what do you think about the default value?

I think it makes sense of turning off `ignoreRestSiblings` by default.
This will match the behavior of ESLint (easier migration).
Also, restructuring and ignoring some properties seems as an edge case to me.
Making it opt-in seems the good call.

File: crates/biome_js_analyze/tests/specs/correctness/noUnusedVariables/invalidRestSiblings.js.snap [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  
+    1 │ // With ignoreRestSiblings: false, unused variables should be reported
+    2 │ const car = { brand: "Tesla", year: 2019, countryCode: "US" };
+  > 3 │ const { brand, year, ...other } = car;
+      │         ^^^^^
+    4 │ console.log(other);
+    5 │ 
+  
+  i Unused variables usually are result of incomplete refactoring, typos and other source of bugs.
+
```
Comment (Conaclos) [inline crates/biome_js_analyze/tests/specs/correctness/noUnusedVariables/invalidRestSiblings.js.snap:34]: I wonder if we could add a note indicating that users can ignore rest sibling with the `ignoreRestSiblings` option.
We could display this note only for rest siblings like in the current case.

**Comment by ematipico** [comment]
@iamakulov are you still interested in this PR? There are few comments to address

**Comment by iamakulov** [comment]
Yes! On it :)

**Comment by iamakulov** [comment]
Done! Again, still no Rust familiarity, so sorry if some PR changes look dumb.

File: crates/biome_js_analyze/tests/specs/correctness/noUnusedVariables/invalidRestSiblings.js.snap (hunk 3 above)
Comment (iamakulov) [reply crates/biome_js_analyze/tests/specs/correctness/noUnusedVariables/invalidRestSiblings.js.snap:33]: Great idea! I added that as a separate note, but let me know if you’d like to have them merged into one.

**Review by ematipico — APPROVED** [review]
I rebased the PR, fixed some grammar, and created a better changeset. Going to merge it
